Adding support for custom port

m4r1vs opened the pull request:
Hi, I just found this beautiful dashboard but was wondering why I wasn't able to set a custom port so I quickly changed a few lines of code and added it to the README.  Now you can specify a port by adding the plugin like this:
```javascript
new Jarvis({
    port: 1337
})
```
this PR should also resolve #32 
I quickly tested that everything works and it falls back to `1337` if no port is given.

imbhargav5 commented:
Hi @m4r1vs 
I am curious about line number 36 in server.js. `${PORT}` would be undefined there, wouldn't it ?

m4r1vs commented:
@imbhargav5  no, I tested it by routing on main to /port (removed in PR) and it returns the port :)

imbhargav5 commented:
@m4r1vs Ah, so it depends on the order of execution and start function executes before the index GET request. Got it. Nevermind! I was about to attempt this PR but you were blazing fast!

m4r1vs commented on src/plugin/index.js:
```diff
@@ -19,7 +21,7 @@ function Jarvis(options) {
 }
 
-Jarvis.prototype.apply = function(compiler) {
+Jarvis.prototype.apply = function (compiler) {
```
Sorry, forgot to disable prettier before making the changes so it added this redundant space here :D

zouhir replied:
Please feel free to use prettier! Just with the native prettier configs would be good!

m4r1vs commented on src/plugin/index.js:
```diff
@@ -6,7 +6,9 @@ const importCwd = require("import-cwd"); // used to get the users project detail
-  // TOOD: add options for port, etc..
+  this.options = {
+    port: options && options.port || 1337 // if no port specified fall back to 1337
+  }
```
And maybe you can add a semicolon here after merging to keep the code-style, I don't use them so I forgot to add one here, sorry

zouhir replied:
I guess I'll have prettier on pre-commit hook but now let's go with default prettier settings from your editor

zouhir commented on src/plugin/index.js:
```diff
@@ -6,7 +6,9 @@ const importCwd = require("import-cwd"); // used to get the users project detail
 const pkg = importCwd("./package.json");
 
 function Jarvis(options) {
```
We can probably use option = {} as a default parameter value then below in the options object no need to check if it's defined. What do you think?

zouhir commented on src/plugin/server.js:
```diff
@@ -31,12 +33,15 @@ if (process.env.NODE_ENV !== "jarvis_dev") {
 } else {
-  app.get("/", (_, res) => res.send("Jarvis client is running on: 1337"));
+  app.get("/", (_, res) => res.send(`Jarvis client is running on: ${PORT}`));
 }
```
The client app uses Socket IO and we probably need to tell it the port isn't 1337 anymore. Is that correct? I hope we don't need to really specify anything on client btw

https://github.com/zouhir/jarvis/blob/master/src/client/components/board.js#L13

m4r1vs replied:
fixed this by adding document.location.port to the socket IO conbection clientside

m4r1vs commented:
Okay, @lukeed and @zouhir, I just fixed that and also added and if statement that checks if specified port is invalid. Port can now also be a string since I added parseInt()

m4r1vs commented:
But if we plan to do server side rendering (which would eat a lot more Resources I think), commit f113d22 needs to be changed because it would throw an error that document is not defined. Something like this:
```javascript
const socket = io("localhost:" 
    + (document === "undefined" && PORT) ? PORT : document.location.port);

ForsakenHarmony commented:
Not real ssr, just injecting the port into the document served is

Sunshine168 commented:
ah  i trying do this too  ,~but seem you already did this . what do you think setting the port with process.env  when build the client 
so in the webpack config  
```
new webpack.DefinePlugin({
  'process.env': {
    'PORT': JSON.stringify(process.env.PORT),
  }
})
```
and with cmd to inject the process.env ~

ForsakenHarmony commented:
That won't work, when you use this it's already bundled

m4r1vs commented:
@Sunshine168 Using process.env.PORT doesn't make sense to me since if you define process.env.PORT it is usually used to define a port for a server serving the bundled application or something and almost never for a dashboard.

reznord commented:
The only option for now is to pass the port as an argument to the plugin itself and then internally use it.

m4r1vs commented:
What exactly won't work in your opinion? @ForsakenHarmony

Sunshine168 commented:
oh , we cant adjust the jarvis client bundle build ?  i mean when we loading the Jarvis plugins, we got the port in option and with cmd  inject the process to build client bundle

m4r1vs commented:
@Sunshine168 I am sorry but I don't quite understand what you mean? The client bundle doesn't get touched by Jarvis, it only observes it.

zouhir commented on src/plugin/index.js:
```diff
@@ -5,17 +5,19 @@ const importCwd = require("import-cwd"); // used to get the users project detail
 
-function Jarvis(options) {
-
+function Jarvis(opts) {
```
sorry I am bit annoying about this, in Node 6 and above we can do this:

`function Jarvis(options = {})` do you think we should do that?

m4r1vs commented:
@zouhir Oups yeah totally, it's 4am right now here in Canada and I can't code when I'm tired, but I will change that real quick..

zouhir commented:
@lukeed : the SSR is next step, that will help us when the people maybe change their hosts file to be `jarvis.team.com` or whatever. but for now, this works fine as a first step!

@Sunshine168 We will choose that setting the port for Jarvis is better done in your Jarvis option, thanks for the contribution to this and your ideas 💜

@m4r1vs I think I am happy with this for now, we might refine as we go and explore our web serving options, but youve been fantastic

Sunshine168 commented:
@m4r1vs  i feel sorry for mine expression , i mean that inject the process.env when Jarvis build Jarvis 's client side  . ah  you may go to take rest

Sunshine168 commented:
@zouhir  this is a Awesome project   expect it will  be stronger   hh

m4r1vs commented:
@Sunshine168 no, it's fine, now I understand what you meant and maybe that could be a solution too later if JARVIS gets build on the client side, and using process.env.PORT can definitely be very useful in some cases!

zouhir commented:
@m4r1vs sorry am being annoying, I tried to run this branch with no `port` in options and the Nodejs server came with 

`RangeError: "port" argument must be >= 0 and < 65536`

I will leave a comment near the line with the only modification need to happen.

Super thanks for your patience.

zouhir commented on src/plugin/index.js:
```diff
@@ -5,8 +5,20 @@ const importCwd = require("import-cwd"); // used to get the users project detail
+function Jarvis(options = {}) {
+  
+  // check if port is valid
+  let portIsValid = true;
```
Since natively, nodeJS server will report an error if the port is not a number or if it's out of range we can remove this bit and delegate that to the server code.

and the function will only become:

```

function Jarvis(options = {}) {
  this.options = {
    port: options.port || 1337 // fall back to 1337 if port is not a number
  };

/* rest of code */
}
```

m4r1vs commented:
I'll do this tomorrow, or later today, depe

m4r1vs commented:
Oh gosh I should really go to bed, I wanted to say that I will do this tomorrow, or later today, depending on time zones 😅

zouhir commented:
haha Go to bed! you worked really hard on this, thank you so much!! we will merge it tomorrow

m4r1vs commented:
I just removed the one part of the if statement, but not sure if this works, if not I'll do that tomorrow, good night or day or whatever :D

zouhir commented:
I think there is still no need to number validate the port 🤫 because number, port range, etc.. will all be reported by nodejs anyway and it's not our concern for now. do you agree?

SUPER thank you for your work on this 💜 this will go in the first feature release and can't wait to have you on contributors board I am designing!

m4r1vs commented:
@zouhir Yeah, I wasn't sure about that and I just added it, but now you said that it is probably done by node/express I am pretty sure you are right. I am currently not at my PC but will push the final changes later today :)

zouhir commented:
@m4r1vs epicccccccccccccccccccccccc sheeeeeeeeeeeeeeeeeeeeeeeetttttttttttttttttttt 💜💜💜💜

m4r1vs commented:
@zouhir commit 9b6aad1 removed manual checking, but I think I should undo that because the error by nodejs when port is a string looks REALLY ugly:
```sh
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: listen EPERM grg
    at Object._errnoException (util.js:1024:11)
    at _exceptionWithHostPort (util.js:1046:20)
    at Server.setupListenHandle [as _listen2] (net.js:1334:19)
    at listenInCluster (net.js:1392:12)
    at Server.listen (net.js:1487:5)
    at Object.start (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack-jarvis/bin/plugin/server.js:43:17)
    at Jarvis.apply (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack-jarvis/bin/plugin/index.js:35:12)
    at Compiler.apply (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/tapable/lib/Tapable.js:375:16)
    at webpack (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack/lib/webpack.js:33:19)
    at processOptions (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack/bin/webpack.js:335:15)
    at yargs.parse (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack/bin/webpack.js:396:2)
    at Object.Yargs.self.parse (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack/node_modules/yargs/yargs.js:533:18)
    at Object.<anonymous> (/mnt/c/Users/mariu/Documents/Projects/attach-developer-web/node_modules/webpack/bin/webpack.js:152:7)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
error Command failed with exit code 1
```

zouhir commented:
UGH, it is ugly :( I just tested this anway and looks great.

1- let's undo that
2- let's pull master into your current branch and fix readme conflict

m4r1vs commented:
@zouhir I just readded that in f40f63a and will resolve conflict..

m4r1vs commented:
Tested this in following scenarios:

1. No options given at all - no error and port 1337 is used
2. Port given, but not a number, and if string, no number in string - error thrown and port 1337 used
3. Port given and is a number or string containing number - given port is used for everything

livingston commented:
Npm does not have the latest version!
https://www.npmjs.com/package/webpack-jarvis

zouhir commented:
@livingston I'm truely sorry, excited and tired, built and forgot to publish 🤣 now it's there!
